### ISSUE TITLE
Can I plz has determinant function?

### ISSUE THREAD

**Issue by alexpeys** [pytorch/pytorch#3423]
Determinants are useful for things like computing optimal designs (for eg. active learning). https://en.wikipedia.org/wiki/Optimal_design 

Can we plz has this?

**Comment by KaiyangZhou** [comment]
@alexpeys You could add a custom torch.autograd.Function ([http://pytorch.org/docs/master/notes/extending.html#extending-torch-autograd](url)) to compute the forward of matrix determinant, as well as the backward to obtain the gradients wrt each element of the matrix.

The codes would look like:

```
class Det(torch.autograd.Function):
    """
    Matrix determinant. Input should be a square matrix
    """

    @staticmethod
    def forward(ctx, x):
        output = torch.potrf(x).diag().prod()**2
        output = torch.Tensor([output]).cuda() # remove .cuda() if you only use cpu
        ctx.save_for_backward(x, output)
        return output

    @staticmethod
    def backward(ctx, grad_output):
        x, output = ctx.saved_variables
        grad_input = None

        if ctx.needs_input_grad[0]:
            grad_input = grad_output * output * x.inverse().t()

        return grad_input
```

To understand how the gradients are computed, please have a look at [https://en.wikipedia.org/wiki/Determinant#Derivative](url)

**Comment by samuela** [comment]
Please please please! Log determinant too.

**Comment by samuela** [comment]
Thanks @KaiyangZhou ! However, that particular approach takes the Cholesky factorization so it requires that the input be PSD. If your googling got you here and you want to do this on general matrices, you'd have to either do `torch.eig` and then massage the complex eigenvalues together or just use numpy.

**Comment by KaiyangZhou** [comment]
@samuela Thank you for pointing this out :)

**Comment by SsnL** [comment]
I'm working on this one. @samuela , do you happen to know the best way to calculate det gradients? There is Jacobi's formula, but I'm not sure how to calculate adjugate efficiently and generally.

**Comment by samuela** [comment]
@SsnL I'm not aware of a method better than Jacobi's formula unfortunately. (Anyone who has more experience with this please chime in!) For all intents and purposes I think it's fair to assume that the input is invertible in which case you can calculate the inverse transpose. Here's how tensorflow and autograd handle it:

* https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/linalg_grad.py#L46
* https://github.com/HIPS/autograd/blob/master/autograd/numpy/linalg.py#L22

It's also derived in section 2.2.4 here: https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf. The log-determinant gradient should be very similar but with a det(A) factor missing.

**Comment by soumith** [comment]
fixed on master via https://github.com/pytorch/pytorch/pull/3816

**Comment by samuela** [comment]
Thanks @soumith and @SsnL!

**Comment by samuela** [comment]
It looks like a PR for this landed on Dec. 1, 2017 but it's not in version 0.3.1. What's the timeline for getting this into a release?

**Comment by SsnL** [comment]
@samuela See https://github.com/pytorch/pytorch/issues/5673 :)

**Comment by samuela** [comment]
@SsnL Ah, gotcha. Looking forward to 0.4 then!
